### PR TITLE
Implement Menu

### PR DESCRIPTION
**Pull request by miguelcobain**
Opening the PR to gather initial feedback on the approach.
Currently opens/closes the popover with internal state.

@NullVoxPopuli is this initial approach what you expected?

Next step is yielding the Item and Separator components from Content.

I think a good first goal for the Menu would be an API like:
```hbs
<Menu as |m|>
  <button type="button" {{m.trigger}}>Menu</button>

  <m.Content as |c|>
    <c.Item>Item 1</c.Item>
    <c.Item>Item 2</c.Item>
    <c.Separator />
    <c.Item>Item 3</c.Item>
  </m.Content>
</Menu>
```

Some questions I have:
- should we allow passing in an `@isOpen` arg, to control the open state of the Menu from the outside?
- if yes to the above, should we provide an accompanying `@onToggle` action?
- in the proposed API, we discussed items and separators as components. Could it make sense to have them as modifiers instead? That would probably more freedom to the caller, no? 🤔 
- `<ComponentSignature @module="components/menu" @name="Signature" />` probably won't work because we're importing `Args` type from another file, right? (i.e `Args: PopoverSignature['Args'];`).
- we should probably default `@placement` arg to `'bottom-start'`, since it's how most menus work like. Do you agree?

### CODE REVIEW

**Comment by NullVoxPopuli** [comment]
> should we allow passing in an @isOpen arg, to control the open state of the Menu from the outside?

I can't think of a use case where we want to force a menu open :thinking: 
if one comes up, there is no reason we can't add it later.

> if yes to the above, should we provide an accompanying @onToggle action?

n/a

> in the proposed API, we discussed items and separators as components. Could it make sense to have them as modifiers instead? That would probably more freedom to the caller, no? 🤔

what would these modifiers do?

> `<ComponentSignature @module="components/menu" @name="Signature" />` probably won't work because we're importing `Args` type from another file, right? (i.e `Args: PopoverSignature['Args'];`).

I'm not sure -- if it ends up not working, we can just say "see Popover args" for now with a link -- the way I'm doing typedoc introspection is super weird, and I have an issue open with the maintainer to maybe one day have some help figuring stuff out :sweat_smile:  github.com/TypeStrong/typedoc/issues/2528

> we should probably default @placement arg to 'bottom-start', since it's how most menus work like. Do you agree?

I do, yes

**Comment by miguelcobain** [comment]
> what would these modifiers do?

They would be used to "register" items and separators. Example:

```hbs
<Menu as |m|>
  <button type="button" {{m.trigger}}>Menu</button>

  <m.Content as |c|>
    <div {{c.item}}>Item 1</div
    <div {{c.item}}>Item 2</div
    <div {{c.separator}}></c.Item>
    <div {{c.item}}>Item 3</div
  </m.Content>
</Menu>
```

`item` and `separator` modifiers would add `role="menuitem"` and `role="separator"` respectively.
Now that I'm reading [aria docs](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/menuitem_role), it feels like using a modifier would get pretty difficult really fast with things like managing `aria-disabled="true"` on the items.

**Comment by NullVoxPopuli** [comment]
Ah yeah, more reason for components for the items and separators.

For the trigger modifier, it can still manage that aria attribute, 'cause i don't think we ever omit it?

**Comment by miguelcobain** [comment]
Another argument is favor is:
> Any button that open a menu should have a role of [button](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role) or, preferably, be a [<button>](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button), and also have aria-haspopup="menu" (or true) set.

So, I think we shouldn't encourage the user to use *any* element at all, since it requires interactivity.
Plus, as mentioned before, a component allows us to manage more things for the user (`aria-controls`, `aria-haspopup`, `aria-expanded`, and probably other things in the future).

**Comment by miguelcobain** [comment]
@NullVoxPopuli this PR is ready for review. 🙏

File: docs-app/public/docs/5-floaty-bits/popover.md [hunk 1]
```diff
@@ -232,4 +232,5 @@ Since a `Popover` isn't an explicit design pattern provided by W3, but instead,
 - [Date Picker Dialog](https://www.w3.org/WAI/ARIA/apg/patterns/dialog-modal/examples/datepicker-dialog/)
 - [Date Picker Combobox](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-datepicker/)
 - [Select-Only Combobox](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-select-only/)
+- [Menu Butto](https://www.w3.org/WAI/ARIA/apg/patterns/menu-button/)
```
Comment (NullVoxPopuli) [inline docs-app/public/docs/5-floaty-bits/popover.md:235]:
```suggestion
- [Menu Button](https://www.w3.org/WAI/ARIA/apg/patterns/menu-button/)
```

File: ember-primitives/package.json [hunk 2]
```diff
@@ -41,9 +41,9 @@
     "@floating-ui/dom": "^1.5.1",
     "decorator-transforms": "^1.1.0",
     "ember-element-helper": "^0.8.4",
-    "ember-velcro": "^2.1.3",
+    "ember-velcro": "^2.2.0",
     "reactiveweb": "^1.2.0",
-    "tabster": "^7.0.1",
+    "tabster": "^7.1.0",
```
Comment (NullVoxPopuli) [inline ember-primitives/package.json:46]: thanks for the upgrade here 🎉

File: ember-primitives/src/components/menu.gts [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+  // focus first focusable element on the content
+  const tabster = getTabster(window);
+  const firstFocusable = tabster?.focusable.findFirst({
+    container: element,
+  });
+
+  firstFocusable?.focus();
+
+  // listen for "outside" clicks
+  function onDocumentClick(e: MouseEvent) {
```
Comment (NullVoxPopuli) [inline ember-primitives/src/components/menu.gts:114]: I'm surprised they/tabster don't have a utility for this!

File: ember-primitives/src/components/menu.gts [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+  >
+    {{yield}}
+  </button>
+</template>;
+
+const IsOpen = () => cell<boolean>(false);
+const TriggerElement = () => cell<HTMLElement>();
+
+export const Menu: TOC<Signature> = <template>
+  {{#let (IsOpen) (uniqueId) (TriggerElement) as |isOpen contentId triggerEl|}}
```
Comment (NullVoxPopuli) [inline ember-primitives/src/components/menu.gts:202]: uniqueId doesn't exist pre 5.1, I think, and in order to support ember-source 4.12, maybe this `Menu` component should be a class component, and we use `guidFor(this)`?

Comment (miguelcobain) [reply ember-primitives/src/components/menu.gts]: Done. 👍

File: ember-primitives/src/components/menu.gts (hunk 3 above)
Comment (miguelcobain) [reply ember-primitives/src/components/menu.gts:116]: Yeah, I didn't find anything for this.

**Comment by miguelcobain** [comment]
@NullVoxPopuli feedback was applied

**Review by NullVoxPopuli — APPROVED** [review]

**Comment by miguelcobain** [comment]
Dependent on https://github.com/CrowdStrike/ember-velcro/issues/190

**Pull request merged**